pallets/gear-voucher/src/lib.rs

```diff
@@ -100,151 +98,413 @@ pub mod pallet {
+        ///                     `None` passed;
+        /// * prolong validity: optionally increases validity block number.
+        #[pallet::call_index(3)]
+        #[pallet::weight(T::WeightInfo::update())]
+        pub fn update(
+            origin: OriginFor<T>,
+            spender: AccountIdOf<T>,
+            voucher_id: VoucherId,
+            move_ownership: Option<AccountIdOf<T>>,
+            balance_top_up: Option<BalanceOf<T>>,
+            append_programs: Option<Option<BTreeSet<ProgramId>>>,
+            prolong_validity: Option<BlockNumberFor<T>>,
+        ) -> DispatchResultWithPostInfo {
+            // Ensuring origin.
+            let origin = ensure_signed(origin)?;
+
+            // Querying voucher.
+            let Some(mut voucher) = Vouchers::<T>::get(spender.clone(), voucher_id) else {
+                return Err(Error::<T>::InexistentVoucher.into());
+            };
+
+            // Ensuring origin is owner of the voucher.
+            ensure!(voucher.owner == origin, Error::<T>::BadOrigin);
+
+            // Flag of extrinsic Noop: if voucher needs update in storage.
+            let mut updated = false;
+
+            // Flattening move ownership back to current owner.
+            let new_owner = move_ownership.and_then(|addr| addr.ne(&voucher.owner).then_some(addr));
+
+            // Optionally updates voucher owner.
+            if let Some(ref owner) = new_owner {
+                voucher.owner = owner.clone();
+                updated = true;
+            }
+
+            // Optionally top ups voucher balance.
+            if let Some(amount) = balance_top_up {
+                if !amount.is_zero() {
+                    T::Currency::transfer(
+                        &origin,
+                        &voucher_id.cast(),
+                        amount,
+                        ExistenceRequirement::AllowDeath,
+                    )
+                    .map_err(|_| Error::<T>::BalanceTransfer)?;
+                    updated = true;
                 }
-                PrepaidCall::SendReply { reply_to_id, .. } => T::Mailbox::peek(who, reply_to_id)
-                    .map(|stored_message| Self::voucher_id(who, &stored_message.source())),
             }
-        }
-    }
-}
 
-#[derive(Debug, Clone, Encode, Decode, TypeInfo, PartialEq, Eq, PartialOrd, Ord)]
-pub enum PrepaidCall<Balance> {
-    SendMessage {
-        destination: ProgramId,
-        payload: Vec<u8>,
-        gas_limit: u64,
-        value: Balance,
-        keep_alive: bool,
-    },
-    SendReply {
-        reply_to_id: MessageId,
-        payload: Vec<u8>,
-        gas_limit: u64,
-        value: Balance,
-        keep_alive: bool,
-    },
-}
+            // Optionally extends whitelisted programs with amount validation.
+            if let Some(extra_programs) = append_programs {
+                if let Some(mut new_programs) = extra_programs {
+                    if let Some(ref mut programs) = voucher.programs {
```

Comment:
are there any security reasons for not setting `programs` if `voucher.programs.is_none()`?

Reply:
> could be the case that some decided to restrict usage of the voucher

we're allowing only extending voucher functionality. if you're about to restrict it - revoke on expiration and issue to user a new one

required by the business logic to support trust-less interactions between issuers and users

Reply:
right, that's more trust-less, agree